src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,503 @@
+        {
+            var cancellationToken = context.CancellationToken;
+            var document = context.Document;
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(context.Span.Start);
+
+            var forStatement = token.Parent.GetAncestorOrThis<TForStatementSyntax>();
+            if (forStatement == null)
+            {
+                return;
+            }
+
+            if (!context.Span.IsEmpty)
+            {
+                // if there is a selection, it must match the 'for' span exactly.
+                if (context.Span != forStatement.GetFirstToken().Span)
+                {
+                    return;
+                }
+            }
+            else
+            {
+                // if there's no selection, defer to the language to decide if it's in an ok location.
+                if (!IsValidCursorPosition(forStatement, context.Span.Start))
+                {
+                    return;
+                }
+            }
+
+            if (!TryGetForStatementComponents(forStatement,
+                    out var iterationVariable, out var initializer,
+                    out var memberAccess, out var stepValueExpressionOpt, cancellationToken))
+            {
+                return;
+            }
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            syntaxFacts.GetPartsOfMemberAccessExpression(memberAccess,
+                out var collectionExpressionNode, out var memberAccessNameNode);
+
+            var collectionExpression = (TExpressionSyntax)collectionExpressionNode;
+            syntaxFacts.GetNameAndArityOfSimpleName(memberAccessNameNode, out var memberAccessName, out _);
+            if (memberAccessName != nameof(Array.Length) && memberAccessName != nameof(IList.Count))
+            {
+                return;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            // If the for-variable is an identifier, then make sure it's declaring a variable
+            // at the for-statement, and not referencing some previously declared symbol.  i.e
+            // VB allows:
+            //
+            //      dim i as integer
+            //      for i = 0 to ...
+            //
+            // We can't convert this as it would change important semantics.
+            // NOTE: we could potentially update this if we saw that the variable was not used
+            // after the for-loop.  But, for now, we'll just be conservative and assume this means
+            // the user wanted the 'i' for some other purpose and we should keep things as is.
+            var iterationSymbol = semanticModel.GetSymbolInfo(iterationVariable.Parent, cancellationToken).GetAnySymbol();
+            if (iterationSymbol != null)
```

Comment:
interesting. so rather than using GetDeclaredSymbol, you use this way.

Reply:
Yup ,you can go either way.  See if it's a declaration and allow if so, or see if it's a reference and disallow if so.  Either seemed fine to me.